tempesta_fw/http.c

```diff
@@ -2756,30 +3344,576 @@ tfw_http_resp_fwd(TfwHttpResp *resp)
+				--name_out->len;
+				TFW_STR_LAST(name_out)->len -= 1;
+				name_found = true;
+			}
+			continue;
+		}
+
+		/*
+		 * LWS is always in the separate chunks between the name and
+		 * value; thus, we can skip length of the entire (LWS) chunks.
+		 */
+		if (!val_found) {
+			if (unlikely(chunk->data[0] == ' '
+				     || chunk->data[0] == '\t'))
+				continue;
+
+			val_out->chunks = chunk;
+			val_found = true;
+		}
+
+		val_out->len += chunk->len;
+
+		/*
+		 * Skip OWS after the header value (RWS); accumulate the length
+		 * in @tail for RWS cutting off (if this is not the end chunk,
+		 * @tail will be reset).
+		 */
+		tail = 0;
+		idx = chunk->len - 1;
+		while (chunk->data[idx] == ' '
+		       || chunk->data[idx] == '\t')
+		{
+			++tail;
+			if (unlikely(!idx))
+				break;
+			--idx;
+		}
+
+		if (unlikely(tail == chunk->len)) {
+			hdr_tail += tail;
+		} else {
+			last_chunk = chunk;
+			last_tail = hdr_tail = tail;
+		}
+	}
+
+	T_DBG3("%s: hdr_tail=%lu, val_out->len=%lu, last_tail=%lu,"
+	       " last_chunk->len=%lu, last_chunk->data='%.*s'\n", __func__,
+	       hdr_tail, val_out->len, last_tail, last_chunk->len,
+	       (int)last_chunk->len, last_chunk->data);
+
+	if (WARN_ON_ONCE(!last_chunk))
+		return -EINVAL;
+
+	val_out->nchunks = chunk - val_out->chunks + 1;
+	val_out->len -= hdr_tail;
+	last_chunk->len -= last_tail;
+
+	return 0;
+}
+
+static int
+tfw_h2_resp_add_loc_hdrs(TfwHttpResp *resp, const TfwHdrMods *h_mods)
+{
+	int r;
+	unsigned int i;
+	TfwHttpTransIter *mit = &resp->mit;
+
+	for (i = 0; i < h_mods->sz; ++i) {
+		const TfwHdrModsDesc *desc = &h_mods->hdrs[i];
+
+		if (test_bit(i, mit->found) || !TFW_STR_CHUNK(desc->hdr, 2))
+			continue;
+
+		r = __hdr_h2_add(resp, desc->hdr);
+		if (unlikely(r))
+			return r;
+	}
+
+	return 0;
+}
+
+/*
+ * Get next header from the @mit->map. Procedure designed to be called from the
+ * outer cycle with changing of @mit iterator (including @mit->curr index of
+ * current header in the indirection map). Note, for optimization purposes, on
+ * each iteration function produces the boundary pointer @mit->bnd for current
+ * iteration and the operation instance @mit->next - for the next iteration
+ * (including source header @mit->next.s_hdr).
+ */
+static int
+tfw_h2_resp_next_hdr(TfwHttpResp *resp, const TfwHdrMods *h_mods)
+{
+	int r;
+	unsigned int i;
+	TfwHttpTransIter *mit = &resp->mit;
+	TfwHttpHdrMap *map = mit->map;
+	TfwNextHdrOp *next = &mit->next;
+	TfwHttpHdrTbl *ht = resp->h_tbl;
+
+	mit->bnd = NULL;
+
+	for (i = mit->curr; i < map->count; ++i) {
+		int k;
+		unsigned short hid = map->index[i].idx;
+		unsigned short d_num = map->index[i].d_idx;
+		TfwStr *tgt = &ht->tbl[hid];
+		TfwStr *first = TFW_STR_CHUNK(tgt, 0);
+		TfwHdrModsDesc *f_desc = NULL;
+
+		tgt = TFW_STR_CHUNK(tgt, d_num);
+		if (WARN_ON_ONCE(!tgt
+				 || TFW_STR_EMPTY(tgt)
+				 || TFW_STR_DUP(tgt)))
+			return -EINVAL;
+
+		if (!h_mods)
+			goto def;
+
+		for (k = 0; k < h_mods->sz; ++k) {
```

Comment:
This function should be treated as a foundation for #1103 (please comment this with `TODO #1103` by the way) and that's good - we can extend `TfwHdrMods` by dynamically decided headers along with always-modified headers (e.g. `Via`) and execute the optimization logic on top of the data structure.

This leads us to typically not so small `h_mods`, but probably that's fine to check the whole list on each processed headers - probably it's cheaper than add additional map to remember which headers we already processed, but please consider the option.

What I don't like at all is the broken idea of HTTP/2 benefits by indexing the headers - we must not run string operations if we can use indexes (see comment for `__hdr_name_cmp()`). HTTP/2 is binary protocol and we decided to make it first class citizen in Tempesta within #309.